docs/src/test-typescript-js.md

```diff
@@ -1,17 +1,109 @@
 ---
 id: test-typescript
-title: "Advanced: TypeScript"
+title: "TypeScript"
 ---
 
+Playwright Test supports TypeScript out of the box. You just write tests in TypeScript and Playwright Test will read them, transform to JavaScript and run. This works both with [CommonJS modules](https://nodejs.org/api/modules.html) and [ECMAScript modules](https://nodejs.org/api/esm.html).
+
+## TypeScript with CommonJS
+
+[Node.js](https://nodejs.org/en/) works with CommonJS modules **by default**. Unless you use `'.mjs'` or `'.mts'` extensions, or specify `type: "module"` in your `pacakge.json`, Playwright Test will treat all TypeScript files as CommonJS. You can then import as usual without an extension.
+
+Consider this helper module written in TypeScript.
+
+```js
+// helper.ts
+export const username = 'John';
+export const password = 'secret';
+```
+
+You can import from the helper as usual.
+
+```js
+// example.spec.ts
+import { test, expect } from '@playwright/test';
+import { username, password } from './helper';
+
+test('example', async ({ page }) => {
+  await page.locator('#username').fill(username);
+  await page.locator('#password').fill(password);
+});
+```
+
+## TypeScript with ESM
+
+You can opt into using [ECMAScript modules](https://nodejs.org/api/esm.html) by setting `type: "module"` in your `package.json` file. Playwright Test will switch to the ESM mode once it reads the `playwright.config.ts` file, so make sure you have one.
+
+Playwright Test [follows the experimental support for ESM in TypeScript](https://www.typescriptlang.org/docs/handbook/esm-node.html) and **requires an extension** when importing from a module, either `'.js'` or `'.ts'`.
```

Comment:
```suggestion
Playwright Test [follows the experimental support for ESM in TypeScript](https://www.typescriptlang.org/docs/handbook/esm-node.html) and (per the spec) **requires an extension** when importing from a module, either `'.js'` or `'.ts'`.
```

to make it more clear that this isn't Playwright quirk, but rather a function of the upstream spec.